File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None :
            return head
        
        # 1. Make a deep of just the list
        curr = head

        while curr:
            copyNode = Node(curr.val)
            copyNode.next = curr.next
            curr.next = copyNode
            curr = curr.next.next

        # 2. Point the random pointers to the copied list
        curr = head
        # print(curr.val)
        while curr:
            if curr.random:
                curr.next.random = curr.random.next
            curr = curr.next.next

        # 3. seperate two list
        curr = head
        currCopy = head.next
        head2 = head.next
  
        while curr:
            curr.next = curr.next.next
            if currCopy.next:
                currCopy.next = currCopy.next.next                
            curr = curr.next
            currCopy = currCopy.next
        # print(head2, head)
        return head2

#Pointers        
# Time Complexity: O(n)
#Space Complexity: O(1)

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head == null){
            return head;
        }

        // 1. Create a deep copy of the original list
        Node curr = head;

        while(curr != null){
            Node currCopy = new Node(curr.val);
            currCopy.next = curr.next;
            curr.next = currCopy;
            curr = curr.next.next;
        }

        // 2. Point random pointers on the copied list same as the original list
        curr = head;

        while(curr != null){
            if(curr.random != null){
                curr.next.random = curr.random.next;
            }
            curr = curr.next.next;
        }

        // 3. Separate original list with copied one

        curr = head;
        Node currCopy = head.next;
        Node head2 = head.next;

        while(curr != null){
            curr.next = curr.next.next;
            if(currCopy.next != null){
                currCopy.next = currCopy.next.next;
            }
            curr = curr.next;
            currCopy = currCopy.next;
        }

        return head2;

    }
}


// #Pointers    Time Complexity: O(n) Space Complexity: O(1)
// 1. Create a deep copy of given list
// 2. allocate the random pointers to new list
// 3. Separate the original list from copied 